### PR TITLE
feature/json workflow bkim

### PR DESCRIPTION
**Pull request by yrkim98**
For issue: https://github.com/AllenCell/napari-aicssegmentation/issues/13

The prototype is finished and I thought I'd throw it out to get some early feedback from everyone before I develop further. Currently still working on more tests and functions, and hooking it up to our MPP to show an example of it running on the plugin.

I created an python object representation of a workflow[(WorkflowEngine)](https://github.com/AllenCell/aics-segmentation/blob/feature/json_workflow_bkim/aicssegmentation/structure_wrapper/WorkflowEngine.py) and a step within a workflow[(WorkflowStep)](https://github.com/AllenCell/aics-segmentation/blob/feature/json_workflow_bkim/aicssegmentation/structure_wrapper/WorkflowStep.py). A lot of my work was based on what @jxchen01 did for his workflow setup functions, but with more added flexibility and functionality. We should be able to string together complex workflows easily using these classes.

Here's a brief example of what it can do:

```
img = np.squeeze(imread(r'C:\Users\brian\Downloads\test_sample_single.tiff'))

# We can load a workflow engine with either a workflow name or a json file path
engine = WorkflowEngine("sec61b", img)
# Can execute all just like before
engine.execute_all()

# We can also execute dynamically
engine = WorkflowEngine("sec61b", img)
print("Executing Step: " + str(engine.currentStep))
print("Step Name: " + str(engine.get_next_step().name))
engine.execute_next()

print("Executing Step: " + str(engine.currentStep))
print("Step Name: " + str(engine.get_next_step().name))
engine.execute_next()

# finished unfinished steps
engine.execute_all()

# We can look at previous results like so
img = engine.get_most_recent_result() # Get result of last step, "Size Filter"
img = engine.get_result(0) # Get result of step 1, Intensity Normalization
img = engine.get_result(2) # Get result of step 3, "2D Filament Filter (slice by slice)"
```

### CODE REVIEW

**Comment by jxchen01** [comment]
@yrkim98  I assume the `.idea` folder is something from your IDE? Do we need to keep them or add to `.gitignore `?

File: aicssegmentation/structure_wrapper/WorkflowStep.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        self.parent: int = step_config["parent"] - 1 # Index of parent in entire workflow #TODO: Better to change json to 0-indexed to avoid confusion
+        self.result: np.ndarray = None               # Result of running this step, None if not executed
+
+        module = importlib.import_module(step_config["module"])
+        self.__function = getattr(module, step_config["function"])
+        self.__parameters: dict = None
+        if "parameter" in step_config:
+            self.__parameters = step_config["parameter"]
+
+    def execute(self, image: np.ndarray) -> np.ndarray:
```
Comment (jxchen01) [inline aicssegmentation/structure_wrapper/WorkflowStep.py:25]: It would be better to use `Union[np.ndarray, List]`. Basically, the input can be a single image or a list of images. 

See https://github.com/AllenCell/aics-segmentation/blob/feature/json_workflow/aicssegmentation/structure_wrapper_config/structure_config_utils.py#L37

to test, you may use this example: https://github.com/AllenCell/aics-segmentation/blob/feature/json_workflow/aicssegmentation/structure_wrapper_config/conf_lmnb1_interphase.json

File: aicssegmentation/structure_wrapper/WorkflowEngine.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        Returns:
+            self.image (np.ndarray): Result of performing workflow step on the given image
+                                     None if step has not been executed yet.
+       """
+        if step_index > self.currentStep:
+            return None # returns None if the WorkflowStep has not been executed.
+        else:
+            return self.steps[step_index].result
+
+    def get_most_recent_result(self) -> np.ndarray:
```
Comment (jxchen01) [inline aicssegmentation/structure_wrapper/WorkflowEngine.py:98]: This is not an issue. But, more like a question. 

If the workflow is nicely sequential, then the concept "next" and "get_most_recent_result" is straightforward. However, for more complicated workflow, I am not sure how to communicate the meaning of "next" and "get_most_recent_result". For example, https://github.com/AllenCell/aics-segmentation/blob/feature/json_workflow/aicssegmentation/structure_wrapper_config/conf_lmnb1_interphase.json

In this example, the output of step 2 goes to step 3, 4, 5, then the output of step 3, 4, 5 are merged in step 6. If you are at step 6, what would be a reasonable "get_most_recent_result"? Currently, it returns step 5 output, which is totally fine in certain context. But, i am not sure this is the best option. I would love to hear more thoughts. 

Also, maybe @thao-do has some thoughts from UI/UX perspective?

**Review by jxchen01 — CHANGES_REQUESTED** [review]

**Review by Aerendel — CHANGES_REQUESTED** [review]
Please add `.idea` to the project .gitignore

File: aicssegmentation/structure_wrapper/WorkflowStep.py (hunk 1 above)
Comment (Aerendel) [reply aicssegmentation/structure_wrapper/WorkflowStep.py:25]: Yes, it looks like we need to accept a List, or a variable number of positional parameters (ie. `*images`)

File: aicssegmentation/tests/test_WorkflowEngine.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+class TestWorkflowEngine:
+    def setup_method(self):
+        self.fake_image = np.zeros([100,100])
+        self.engine = WorkflowEngine("sec61b", self.fake_image)
+
+    def test_get_next_step(self):
+        assert self.engine.get_next_step().name == "Intensity Normalization"
+        self.engine.execute_next()
+        assert self.engine.get_next_step().name == "Edge Preserving Smoothing"
+        self.engine = WorkflowEngine("sec61b", self.fake_image)
```
Comment (schoinh) [inline aicssegmentation/tests/test_WorkflowEngine.py:15]: Why is this line necessary? 🤔

File: aicssegmentation/tests/test_WorkflowEngine.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+    def test_execute_next(self):
+        assert self.engine.currentStep == 0
+        assert self.engine.steps[0].result is None
+        self.engine.execute_next()
```
Comment (schoinh) [inline aicssegmentation/tests/test_WorkflowEngine.py:20]: This might be a matter of preference but I find it a bit confusing that we're calling a function called `execute_next()` but we're really executing `currentStep` not `nextStep`. Not sure what the best way to fix this would be, maybe renaming `execute_next()` as `execute_step()`? Or renaming `currentStep` as `nextStep`?

(Also in ln 20 of WorkflowEngine.py your comment says "next step to execute" but the property name is `self.currentStep` 😛)

File: aicssegmentation/structure_wrapper/WorkflowEngine.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+            # First image, so use the starting image for the next workflow step
+            image = self.starting_image
+        elif self.isDone():
+            # No more workflow steps to perform
+            # TODO: what to do if done with workflow but execute_next is prompted?
+            # printing message for now
+            print("No steps left to run")
+        else:
+            # First step has been run, so run next workflow step with the result of its parent
+            image = self.get_result(self.get_next_step().parent)
```
Comment (benjijamorris) [inline aicssegmentation/structure_wrapper/WorkflowEngine.py:72]: I don't know if this is something that should be solved in the engine or in how the segmentation functions are implemented, but right now we have some steps that return multiple images [(e.g Masked Object Thresholding)](https://github.com/AllenCell/aics-segmentation/blob/feature/json_workflow/aicssegmentation/core/MO_threshold.py) and subsequent steps that only rely on one of those images [(e.g.  NPM1)](https://github.com/AllenCell/aics-segmentation/blob/feature/json_workflow/aicssegmentation/structure_wrapper_config/conf_npm1.json) @jxchen01 should I split the MO threshold into two functions to avoid this problem?

Comment (jxchen01) [reply aicssegmentation/structure_wrapper/WorkflowEngine.py:72]: i will take a look. Thanks for bringing this up

File: aicssegmentation/structure_wrapper/WorkflowStep.py (hunk 1 above)
Comment (yrkim98) [reply aicssegmentation/structure_wrapper/WorkflowStep.py:25]: When it is a list of images, do I have to execute the function on each image on the list or will it be fine to just pass it to your `aics-segmentation` functions?

File: aicssegmentation/structure_wrapper/WorkflowEngine.py (hunk 2 above)
Comment (yrkim98) [reply aicssegmentation/structure_wrapper/WorkflowEngine.py:98]: This is also a thought I had while creating this part of the class. You're right that currently if we call `get_most_recent_result()`, it will just grab the image that has most recently been executed. 

To allow for more functionality in the future i also created a `get_result(int: step_index)` function. Given the index of a step in the workflow, we can also grab results that way.

I realize that we may not have perfectly sequential workflows, but if we sort the `WorkflowSteps` list in the `WorkflowEngine` by parent id, and execute each step sequentially, we can still execute complicated workflows without any problems. For the example you provided above, if we have the result of 3, 4, and 5, then call step 6, the results from the previous steps are available to merge in step 6.

I would love to hear @thao-do 's thoughts from the UI perspective as well!

File: aicssegmentation/tests/test_WorkflowEngine.py (hunk 3 above)
Comment (yrkim98) [reply aicssegmentation/tests/test_WorkflowEngine.py:15]: I'm "resetting" the workflow engine to a new instance so I can test from the start again.

File: aicssegmentation/tests/test_WorkflowEngine.py (hunk 4 above)
Comment (yrkim98) [reply aicssegmentation/tests/test_WorkflowEngine.py:20]: Yeah this is a good point! Definitely a little confusing. For clarity I'm renaming currentStep to nextStep, since that's what that value represents.

File: aicssegmentation/tests/test_WorkflowEngine.py (hunk 3 above)
Comment (schoinh) [reply aicssegmentation/tests/test_WorkflowEngine.py:15]: Ok, this is totally nit picking now I know... but from what I'm reading it looks like `setup_method` should run at the beginning of every method in this class, so it shouldn't be necessary.

File: aicssegmentation/structure_wrapper/WorkflowEngine.py (hunk 5 above)
Comment (jxchen01) [reply aicssegmentation/structure_wrapper/WorkflowEngine.py:72]: @benjijamorris The second output of MO is only for debugging. So, i think at least for MO, we can return a single output. Are you aware of any other function that returns multiple outputs?

Comment (benjijamorris) [reply aicssegmentation/structure_wrapper/WorkflowEngine.py:72]: I think that's the only one we're using now

File: aicssegmentation/tests/test_WorkflowEngine.py (hunk 3 above)
Comment (Aerendel) [reply aicssegmentation/tests/test_WorkflowEngine.py:15]: Correct, no need to bother resetting anything that's defined by the `setup_method` at the end of a test.

Comment (yrkim98) [reply aicssegmentation/tests/test_WorkflowEngine.py:15]: Ah I did not know that! I'll make that fix

**Review by heeler — COMMENTED** [review]
I really like the typing you're doing on functions. Generally, this looks good. The one thing I would encourage is to incorporate `mypy` to do strict type checking. It's a good way to catch exceptions. You can override it in cases where the rigor is too constraining.

File: aicssegmentation/structure_wrapper/WorkflowEngine.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        Params:
+            workflow_name (str): dictionary object containing information about this workflow step
+            image (np.ndarray):  image to perform workflow on
+        """
+        self.workflow_name: str = workflow_name   # Workflow name
+        self.steps: list = self.__get_steps()        # List of WorkflowSteps for this workflow
+        self.currentStep: int = 0                 # Next step to execute
+        self.starting_image: np.ndarray = image          # Initial image
+
+    def __get_steps(self) -> list:
```
Comment (Aerendel) [inline aicssegmentation/structure_wrapper/WorkflowEngine.py:23]: Correct type annotation for a list is `List` (from `typing.List`). Lowercase `list` is a function that creates a list.

File: aicssegmentation/structure_wrapper/WorkflowEngine.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+            none
+
+        Returns:
+            (np.ndarray): Result of the final WorkflowStep.
+       """
+        while not self.isDone():
+            self.execute_next()
+        return self.get_most_recent_result()
+
+    def isDone(self) -> bool:
```
Comment (Aerendel) [inline aicssegmentation/structure_wrapper/WorkflowEngine.py:128]: A wild camelCase appears :laughing: 
Change to `is_done` to be consistent?

File: aicssegmentation/structure_wrapper/WorkflowStep.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
+import importlib
+import numpy as np
+
+class WorkflowStep:
+    """
+    A class that defines a step in an AICS-Segmentation workflow.
+    """
+    def __init__(self, step_config: dict):
```
Comment (Aerendel) [inline aicssegmentation/structure_wrapper/WorkflowStep.py:8]: Correct type annotation for a list is `Dict` (from `typing.Dict`). Lowercase `dict` is a function that creates a Dictionary.

File: aicssegmentation/structure_wrapper_config/structure_config_utils.py [hunk 9]
```diff
@@ -3,25 +3,30 @@
...
 
 def load_workflow_config(workflow_name: str):
-
     json_path = Path(__file__).parent / f"conf_{workflow_name}.json"
     with open(json_path, "r") as read_file:
         cfg = json.load(read_file)
 
     return cfg
 
+def parse_config_to_objects(cfg: dict):
```
Comment (Aerendel) [inline aicssegmentation/structure_wrapper_config/structure_config_utils.py:25]: Uppercase `Dict`

**Review by Aerendel — CHANGES_REQUESTED** [review]

**Comment by heeler** [comment]
> @heeler Does it make sense to setup a `mypy` check in the Segmenter CICD? I haven't actually used it extensively before so not sure how flexible it is (you mention it might be too constraining in some cases?). I am personally all for 100% strict typing if possible. Wouldn't set it up for this repo though since we'd have to go and annotate all of Jianxu's code.

I thought about this but it depends on how much work it is to make the code Na Hyung and Brian have implemented compliant. I don' think it would require the aicssegmentation module to be strickly typed though at least having typing on the interface functions would be good if possible. Adding it to the GH Actions would not be hard (we've integrated it in aicsimageio 4.0 main branch already) but I would suggest Brian and Na Hyung play around with it first. It is a good way to catch bugs but it does still have awkward spots.

**Comment by yrkim98** [comment]
> Please add `.idea` to the project .gitignore

> @yrkim98 I assume the `.idea` folder is something from your IDE? Do we need to keep them or add to `.gitignore `?

I've already added `.idea/` to my `.gitignore` file. Any idea why the changes are still being tracked?

File: aicssegmentation/structure_wrapper/WorkflowStep.py (hunk 1 above)
Comment (yrkim98) [reply aicssegmentation/structure_wrapper/WorkflowStep.py:25]: @jxchen01 Still curious about my above question, please let me know!

File: aicssegmentation/structure_wrapper/WorkflowEngine.py [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+        Params:
+            workflow_name (str): dictionary object containing information about this workflow step
+            image (np.ndarray):  image to perform workflow on
+        """
+        self.workflow_name: str = workflow_name   # Workflow name
+        self.steps: list = self.__get_steps()        # List of WorkflowSteps for this workflow
+        self.next_step: int = 0                 # Next step to execute
+        self.starting_image: np.ndarray = image          # Initial image
+
+    def __get_steps(self) -> List:
```
Comment (Aerendel) [inline aicssegmentation/structure_wrapper/WorkflowEngine.py:24]: Full typing annotation for a typed list would be `List[WorkflowStep]`

Also, I would rename this method to something more explicit like `__load_steps_from_config` because we're going to need a public `get_steps` method or a `steps` property (that I don't see currently implemented) that the plugin would use to then display the steps on the UI.

File: aicssegmentation/structure_wrapper/WorkflowStep.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+            image (np.ndarray): Image to perform this workflow step on, generally parent image
+
+        Returns:
+            self.result (np.ndarray): Result of performing workflow step on the given image.
+       """
+        if self.__parameters:
+            self.result: np.ndarray = self.__function(*image, **self.__parameters)
+        else:
+            # for "Merge Segmentation' we need a list of images, while other functions take in with multiple params
+            if self.name == "Merge Segmentation":
```
Comment (Aerendel) [inline aicssegmentation/structure_wrapper/WorkflowStep.py:47]: I'm not a huge fan of hard-coding some logic based on the name of a step. Can we come up with something more explicit / robust, for example a property on the WorkflowStep to mark whatever is going on here?

**Review by Aerendel — CHANGES_REQUESTED** [review]
Tests for WorkflowStep need to be written

File: aicssegmentation/structure_wrapper/WorkflowStep.py (hunk 11 above)
Comment (yrkim98) [reply aicssegmentation/structure_wrapper/WorkflowStep.py:47]: Hm, I wonder if there is a way to get the function we're calling, and see what it expects as parameters. This was the single exception where we need a list of images so I thought hard coded it for now. Will look into a cleaner solution

Comment (jxchen01) [reply aicssegmentation/structure_wrapper/WorkflowStep.py:47]: our current solution on this issue is something like this: https://github.com/AllenCell/aics-segmentation/blob/feature/json_workflow/aicssegmentation/structure_wrapper_config/structure_config_utils.py#L37

basically, just check if the parent is a list or not. 

Also, "Merge Segmentation" is not the only one that takes mulitple inputs. There are quite a few more. For example https://github.com/AllenCell/aics-segmentation/blob/feature/json_workflow/aicssegmentation/structure_wrapper_config/conf_lmnb1_interphase.json#L78

Comment (yrkim98) [reply aicssegmentation/structure_wrapper/WorkflowStep.py:47]: My code already supports multiple inputs in function parameters. I think my issue is different- I need to know whether a function expects multiple images as one list(eg. the `segmentation_union()` function) or multiple images unpacked into individual parameters (eg. the `get_seed_for_objects()` function) I tried your code on the lmnb1_interphase workflow and still got this error on `segmentation_union()`
image

Comment (yrkim98) [reply aicssegmentation/structure_wrapper/WorkflowStep.py]: Would this be a good place to use a try/except? This would work:
```
            try:
                # Most functions require unpacking the images into individual params
                self.result: np.ndarray = self.__function(*image)
            except:
                # Some functions want it as a list
                self.result: np.ndarray = self.__function(image)
```

**Comment by yrkim98** [comment]
All workflows should now work given a valid json file. I can run 21/25 workflows using the engine
The following workflows do not work:

- lamp1 : empty json file
- lmb1_mitotic : empty json file
- npm1 : missing a fourth step in the json file
- son: empty json file

**Comment by jxchen01** [comment]
> All workflows should now work given a valid json file. I can run 21/25 workflows using the engine
> The following workflows do not work:
> 
> * lamp1 : empty json file
> * lmb1_mitotic : empty json file
> * npm1 : missing a fourth step in the json file
> * son: empty json file

i am working those four config files now and should be done very soon

File: aicssegmentation/structure_wrapper/WorkflowStep.py (hunk 11 above)
Comment (jxchen01) [reply aicssegmentation/structure_wrapper/WorkflowStep.py]: looks ok to me

**Comment by jxchen01** [comment]
@yrkim98 All four workflows you mentioned above should have been fixed. Could you try again?

File: aicssegmentation/structure_wrapper/WorkflowStep.py (hunk 11 above)
Comment (Aerendel) [reply aicssegmentation/structure_wrapper/WorkflowStep.py]: I think that's an ok workaround for now. My question would be: could we instead standardize the functions so that they all either take a List of images, or as multiple params - but not both? If all functions take images as inputs, it should make sense to standardize the input format to stay consistent.

File: aicssegmentation/structure_wrapper/WorkflowStep.py [hunk 12]
```diff
@@ -43,11 +43,10 @@ def execute(self, image: List) -> np.ndarray:
...
-            # for "Merge Segmentation' we need a list of images, while other functions take in with multiple params
-            if self.name == "Merge Segmentation":
-                # feed in as list
-                self.result: np.ndarray = self.__function(image)
-            else:
-                # unpack list and feed in images
+            try:
+                # Most functions require unpacking the images into individual params
                 self.result: np.ndarray = self.__function(*image)
+            except:
```
Comment (Aerendel) [inline aicssegmentation/structure_wrapper/WorkflowStep.py:49]: If we're sticking with this version - we should only catch the proper exception instead of a blanket `except`.
So something like this:
`except Typerror:`

**Review by Aerendel — CHANGES_REQUESTED** [review]

File: aicssegmentation/structure_wrapper/WorkflowStep.py [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+       """
+
+        if self.__parameters:
+            self.result: np.ndarray = self.__function(*image,
+                                                      **self.__parameters)
+        else:
+            try:
+                # Most functions require unpacking the images
+                self.result: np.ndarray = self.__function(*image)
+            except Exception:
```
Comment (Aerendel) [inline aicssegmentation/structure_wrapper/WorkflowStep.py:70]: You need to actually catch the proper error type that is thrown there. I believe it's `TypeError` but you'll need to verify.

**Review by Aerendel — CHANGES_REQUESTED** [review]

File: aicssegmentation/structure_wrapper/WorkflowStep.py [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+                                        on the given image.
+        """
+
+        if self.__parameters:
+            self.result: np.ndarray = self.__function(*image, **self.__parameters)
+        else:
+            try:
+                # Most functions require unpacking the images
+                self.result: np.ndarray = self.__function(*image)
+            except KeyError:
```
Comment (Aerendel) [inline aicssegmentation/structure_wrapper/WorkflowStep.py:69]: Let's get a unit test for this, and unit tests for `WorkflowStep` in general

**Review by Aerendel — CHANGES_REQUESTED** [review]

File: aicssegmentation/tests/test_WorkflowStep.py [hunk 15]
```diff
@@ -0,0 +1,47 @@
+import numpy as np
+from aicssegmentation.structure_wrapper.WorkflowEngine import WorkflowStep
+from skimage import data
+import pytest
+
+
+class TestWorkflowEngine:
```
Comment (Aerendel) [inline aicssegmentation/tests/test_WorkflowStep.py:7]: Bad copy paste - this should be `TestWorkflowStep`. Not sure if pytest is even executing this right now since there are 2 test classes with the same name.

**Review by Aerendel — CHANGES_REQUESTED** [review]

File: aicssegmentation/tests/test_WorkflowStep.py [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+    def setup_method(self):
+        caller = getattr(data, "astronaut")
+        image = caller()
+        self.fake_image = np.asarray(image)
+
+        self.test_dict = dict()
+        self.test_dict["name"] = "intensity_normalization"
```
Comment (Aerendel) [inline aicssegmentation/tests/test_WorkflowStep.py:14]: Since you're effectively creating a new WorkflowStep in each test, creating one in your setup is somewhat useless. Not a big deal but I think it makes the test file more confusing.

**Review by Aerendel — APPROVED** [review]
